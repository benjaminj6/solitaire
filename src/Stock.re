let component = ReasonReact.statelessComponent("Stock");

let make = (~cards, ~onClick, _children) => {
  ...component,
  render: _self =>
    switch (cards) {
    | [] => <EmptyCard onClick=(_ev => onClick()) />
    | [card, ..._rest] =>
      <Card
        card={...card, selectable: true}
        onClick=((~card as _c) => onClick())
      />
    },
};