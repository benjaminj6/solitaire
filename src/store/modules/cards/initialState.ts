import * as range from 'lodash/fp/range';
import { Card, CardRank, CardRankNumber, Deck, State, Suit } from './types';

type GetRankFromRemainder = (x: CardRankNumber) => CardRank;
export const getRankFromRemainder: GetRankFromRemainder = num => {
  switch (num) {
    case '0':
      return 'K';
    case '1':
      return 'Q';
    case '2':
      return 'J';
    case '12':
      return 'A';
    default:
      return `${13 - +num}` as CardRank;
  }
};

type GetInitialSuit = (x: Suit) => Deck;
export const getInitialSuit: GetInitialSuit = suitName => {
  return range(1, 14)
    .map((num: number) => {
      const value = `${num % 13}` as CardRankNumber;
      const rank = getRankFromRemainder(value);

      return {
        id: `${suitName}-${rank}`,
        rank,
        suit: suitName,
        value: 13 - +value,
        visible: false,
      };
    })
    .sort((first: Card, second: Card) => first.value - second.value)
    .reduce((obj: Deck, card: Card) => {
      obj[card.id] = card;
      return obj;
    }, {});
};

type GetInitialDeck = () => Deck;
export const getInitialDeck: GetInitialDeck = () => {
  const suits = ['hearts', 'spades', 'diamonds', 'clubs'];
  return suits.reduce(
    (obj: Deck, suitName: Suit) => ({
      ...obj,
      ...getInitialSuit(suitName),
    }),
    {}
  );
};

export default {
  deck: getInitialDeck(),
  shuffledDeck: [],
} as State;
