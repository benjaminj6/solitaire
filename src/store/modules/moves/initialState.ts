import { State } from './types';

const initialState: State = {
  nextMove: {},
  moves: {},
  history: [],
};

export default initialState;