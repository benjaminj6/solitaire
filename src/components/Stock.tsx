import * as React from 'react';
import { Deck, Stock } from '../store/modules/cards/types';
import Card from './Card';
import EmptyCard from './EmptyCard';
import { last } from 'lodash';

type Props = {
  className?: string;
  style?: { [key: string]: string };
  deck: Deck;
  stock: Stock;
  onClickCard?: React.EventHandler<React.MouseEvent<any>>;
  onClickEmpty?: React.EventHandler<React.MouseEvent<any>>;
};

const StockComponent = (props: Props) => (
  <React.Fragment>
    {props.stock.length ? (
      <Card
        {...props.deck[last(props.stock)]}
        className="h-full w-full"
        onClick={props.onClickCard}
      />
    ) : (
      <EmptyCard onClick={props.onClickEmpty} />
    )}
  </React.Fragment>
);

export default StockComponent;
