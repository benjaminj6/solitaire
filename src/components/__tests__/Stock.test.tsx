import * as React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import Stock from '../Stock';
import initialState from '../../store/modules/cards/initialState';
import {
  mockConsoleError,
  restoreConsoleError,
} from '../../../test-utils/__mocks__/consoleError';

const renderer = createRenderer();

beforeAll(mockConsoleError);
afterAll(restoreConsoleError);

describe('Stock', () => {
  test('should render', () => {
    const tree = renderer.render(<Stock
      deck={initialState.deck}
      stock={['hearts-2', 'spades-K', 'diamonds-4', 'hearts-A']}
    />);

    expect(tree).toMatchSnapshot();
  });

  test('should render with an empty card if there is no length in the stock', () => {
    const tree = renderer.render(<Stock deck={initialState.deck} stock={[]} />);
  });
});
