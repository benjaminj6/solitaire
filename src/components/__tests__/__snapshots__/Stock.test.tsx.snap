// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stock should render 1`] = `
<UNDEFINED>
  <Card
    className="h-full w-full"
    id="hearts-A"
    rank="A"
    suit="hearts"
    value={1}
    visible={false}
  />
</UNDEFINED>
`;
